File: sandbox/scripts/live-stream-test.mjs
import { CodeLanguage, Sandbox } from '../dist/src/index.js';

const apiKey = process.env.BRIMBLE_SANDBOX_KEY;
if (!apiKey) {
  console.error('Set BRIMBLE_SANDBOX_KEY to run this script.');
  process.exit(1);
}

async function readFileStream(stream) {
  const decoder = new TextDecoder();
  let text = '';

  for await (const chunk of stream) {
    text += decoder.decode(chunk, { stream: true });
  }

  return text + decoder.decode();
}

const client = new Sandbox({ apiKey, timeoutMs: 120_000 });

console.log('Creating sandbox...');
const sandbox = await client.sandboxes.createReady({ template: 'node-22' });
console.log(`Sandbox ready: ${sandbox.id}\n`);

let failed = false;

function pass(label) {
  console.log(`PASS  ${label}`);
}

function fail(label, error) {
  failed = true;
  console.error(`FAIL  ${label}`);
  console.error(error);
}

try {
  console.log('1) Buffered exec');
  const buffered = await sandbox.exec({ cmd: 'echo hello-buffered' });
  if (buffered.exit_code === 0 && buffered.stdout.includes('hello-buffered')) {
    pass('buffered exec');
  } else {
    fail('buffered exec', buffered);
  }

  console.log('\n2) Streaming exec with for await');
  const output = await sandbox.exec({
    cmd: 'for i in 1 2 3; do echo line-$i; sleep 0.15; done',
    stream: true,
  });

  const logs = [];
  for await (const log of output) {
    logs.push(log);
    process.stdout.write(`   [${log.stream}] ${JSON.stringify(log.data)}`);
  }
  process.stdout.write('\n');

  const streamed = await output.result();
  if (logs.length >= 1 && streamed.exit_code === 0 && streamed.stdout.includes('line-1')) {
    pass('streaming exec');
  } else {
    fail('streaming exec', { logs, streamed });
  }

  console.log('\n3) Callback streaming exec');
  const callbackChunks = [];
  const callbackResult = await sandbox.exec({
    cmd: 'printf callback-ok',
    onStdout: (chunk) => {
      callbackChunks.push(chunk);
      process.stdout.write(`   [callback] ${JSON.stringify(chunk)}`);
    },
  });
  process.stdout.write('\n');

  if (callbackChunks.length > 0 && callbackResult.stdout.includes('callback-ok')) {
    pass('callback exec');
  } else {
    fail('callback exec', { callbackChunks, callbackResult });
  }

  console.log('\n4) Streaming runCode');
  const codeOutput = await sandbox.runCode({
    language: CodeLanguage.Node,
    code: 'for (let i = 1; i <= 2; i++) console.log(`code-${i}`)',
    stream: true,
  });

  const codeLogs = [];
  for await (const log of codeOutput) {
    codeLogs.push(log);
  }

  const codeResult = await codeOutput.result();
  if (codeLogs.length >= 1 && codeResult.exit_code === 0 && codeResult.stdout.includes('code-1')) {
    pass('streaming runCode');
  } else {
    fail('streaming runCode', { codeLogs, codeResult });
  }

  console.log('\n5) File upload + streamed download');
  await sandbox.putFile('tmp/live-stream-test.txt', Buffer.from('local-package-stream-ok', 'utf-8'));
  const fileStream = await sandbox.getFile('tmp/live-stream-test.txt');
  const fileContents = await readFileStream(fileStream);

  if (fileContents.includes('local-package-stream-ok')) {
    pass('file stream download');
  } else {
    fail('file stream download', { fileContents });
  }
} catch (error) {
  failed = true;
  console.error('\nUnexpected error:', error);
} finally {
  console.log('\nDestroying sandbox...');
  try {
    await sandbox.destroy();
    console.log('Sandbox destroyed.');
  } catch (error) {
    console.error('Cleanup failed:', error);
  }
}

if (failed) {
  process.exit(1);
}

console.log('\nAll live streaming checks passed.');

File: sandbox-python/scripts/live_stream_test.py
#!/usr/bin/env python3
"""Live streaming test against sandbox.brimble.io using the local package."""

from __future__ import annotations

import os
import sys

from brimble_sandbox import Sandbox
from brimble_sandbox.enums import CodeLanguage

api_key = os.environ.get("BRIMBLE_SANDBOX_KEY")
if not api_key:
    print("Set BRIMBLE_SANDBOX_KEY")
    sys.exit(1)

client = Sandbox(api_key=api_key)
failed = False

print("Creating sandbox...")
sandbox = client.sandboxes.create_ready({"template": "node-22"})
print(f"Sandbox ready: {sandbox.id}\n")

try:
    print("1) Buffered exec")
    buffered = sandbox.exec({"cmd": "echo hello-buffered"})
    if buffered["exit_code"] == 0 and "hello-buffered" in buffered["stdout"]:
        print("PASS  buffered exec")
    else:
        failed = True
        print("FAIL  buffered exec", buffered)

    print("\n2) Streaming exec (for log in output)")
    output = sandbox.exec_stream({"cmd": "for i in 1 2 3; do echo line-$i; sleep 0.15; done"})
    logs = list(output)
    result = output.result()
    for log in logs:
        print(f"   [{log['stream']}] {log['data']!r}")
    if logs and result["exit_code"] == 0 and "line-1" in result["stdout"]:
        print("PASS  streaming exec")
    else:
        failed = True
        print("FAIL  streaming exec", logs, result)

    print("\n3) Callback exec")
    chunks: list[str] = []
    callback = sandbox.exec(
        {"cmd": "printf callback-ok"},
        on_stdout=chunks.append,
    )
    print(f"   [callback] {chunks!r}")
    if chunks and "callback-ok" in callback["stdout"]:
        print("PASS  callback exec")
    else:
        failed = True
        print("FAIL  callback exec", chunks, callback)

    print("\n4) Streaming run_code")
    code_output = sandbox.run_code_stream(
        {
            "language": CodeLanguage.NODE,
            "code": "for (let i = 1; i <= 2; i++) console.log(`code-${i}`)",
        }
    )
    code_logs = list(code_output)
    code_result = code_output.result()
    if code_logs and code_result["exit_code"] == 0 and "code-1" in code_result["stdout"]:
        print("PASS  streaming run_code")
    else:
        failed = True
        print("FAIL  streaming run_code", code_logs, code_result)

    print("\n5) File upload + streamed download")
    sandbox.put_file("tmp/live-python-stream.txt", b"local-python-stream-ok")
    file_stream = sandbox.get_file("tmp/live-python-stream.txt")
    file_contents = b"".join(file_stream).decode("utf-8")
    file_stream.close()
    if "local-python-stream-ok" in file_contents:
        print("PASS  file stream download")
    else:
        failed = True
        print("FAIL  file stream download", file_contents)

finally:
    print("\nDestroying sandbox...")
    sandbox.destroy()
    print("Sandbox destroyed.")

if failed:
    sys.exit(1)

print("\nAll Python live streaming checks passed.")
